base/linalg/diagonal.jl

```diff
@@ -180,8 +180,12 @@ end
 (\)(D::Diagonal, b::AbstractVector) = reshape(scale(1 ./ D.diag, reshape(b, length(b), 1)), length(b))
 (\)(Da::Diagonal, Db::Diagonal) = Diagonal(Db.diag ./ Da.diag)
 
+invsimilar{T}(d::Vector{T}) = similar(d, T)
+invsimilar{T<:Integer}(d::Vector{T}) = similar(d, Float64)
+invsimilar{T<:Integer}(d::Vector{Complex{T}}) = similar(d, Complex{Float64})
+
 function inv{T}(D::Diagonal{T})
-    Di = similar(D.diag)
+    Di = invsimilar(D.diag)
```

Comment:
@andreasnoack Nice idea! I thought `typeof(inv(one(T)))` is enough and fixed my commit.